### DISCUSSION TITLE
How to prevent HTTP 429: Too many requests

### DISCUSSION THREAD

**Discussion by martinleppen** [jolpica/jolpica-f1#80] [Q&A] [answered]
I don't know if this is the correct place to ask this question, but I'm a Mendix developer with a hobby project/application where I made use of the Ergast API. I changed the URL to this new version, but for the lap times calls for a race, I'm getting a 429: Too many requests response.
For other calls I increased the limit, so the total amount of calls wouldn't be so high If I want to collect all the data, but for laps I have to create a call for each of the laps.
Should I implement my call different, or is there something else I can do to prevent this response?
Thanks for the help

**Comment by jolpica** [comment, marked as answer]
Hey Martin,

The error occurs because you've hit our rate limit of 4 per second, or 500 per hour (This will likely decrease in the future once we set up self-service API Tokens).
If most of your requests are getting 429s its likely because you've hit the hourly limit.

For everyone who may read this, there are 2 options:
1. Reduce your calls to the API to below the sustained and burst rates.
2. If you believe you have a valid use case, request an API Token with a higher rate limit.

In your case I suspect option 2 isn't required and instead more efficient requests / caching would be best.

What data are you trying to retrieve? What api requests are you making?

**Reply by martinleppen** [reply]
Thanks for the quick response. I wanted to update the missing data I hadn't obtained yet, since I encountered issues with the Ergast API at the start of the season.
Checking my logs, I can see that the 4 per second was indeed the rate limit. Because I'm implementing some statistics analysis (like lap time comparison between drivers), I want to have all the data in my own application, so that's the reason for all of the calls.
I will implement functionality to keep track of the amount of calls and the speed of the calls, to prevent further issues.

**Reply by jolpica** [reply]
Are you making calls similar to these?
`/ergast/f1/2024/17/laps/35.json?offset=0&limit=100`
`/ergast/f1/2024/17/laps/36.json?offset=0&limit=100`


If so I'd recommend removing the filter at the end and use paging instead. 
e.g. instead of `api.jolpi.ca/ergast/f1/2024/17/laps/9.json` from 1-70 (or however many laps).
Do `api.jolpi.ca/ergast/f1/2024/17/laps.json?limit=100?offset=0` for offset of 0,100,etc

This should reduce the number of requests you need to make, and get you the data you need quicker. (Will also help out the API a bit too 😅 )

**Reply by martinleppen** [reply]
Yes, I was making the call per lap. Didn't know I could remove that filter. I'll update my implementation to be able to handle receiving not all the lap times of a single lap in one request. Thanks for the help.
